addons/website/models/website.py

```diff
@@ -1340,3 +1341,235 @@ def _get_cached_values(self):
+            search_detail['results_data'] = results_data
+        return search_details
+
+    def _search_find_fuzzy_term(self, search_details, search, limit=1000, word_list=None):
+        """
+        Returns the "closest" match of the search parameter within available words.
+
+        :param search_details: obtained from `_search_get_details()`
+        :param search: search term to which words must be matched against
+        :param limit: maximum number of records fetched per model to build the word list
+        :param word_list: if specified, this list of words is used as possible targets instead of
+            the words contained in the match fields of each involved model
+
+        :return: term on which a search can be performed instead of the initial search
+        """
+        def trigram_enumerate_words():
+            """
+            Browses through all words that need to be compared to the search term.
+            It extracts all words of every field associated to models in the fields_per_model parameter.
+            The search is restricted to a records having the non-zero pg_trgm.word_similarity() score.
+            """
+            match_pattern = '\\w{%s,}' % min(4, len(search) - 3)
+            search_term = "'%s'" % escape_psql(search)
+            self.env.cr.execute("SET pg_trgm.word_similarity_threshold TO 0.3")
+            for search_detail in search_details:
+                model_name, fields = search_detail['model'], search_detail['search_fields']
+                model = self.env[model_name]
+                if search_detail.get('requires_sudo'):
+                    model = model.sudo()
+                fields = ['arch_db' if field == 'arch' else field for field in fields if field in model]
+                if model_name == 'website.page':
+                    fields.remove('name') # TODO handle ?
+
+                if self.env.registry.has_unaccent:
+                    unaccent = get_unaccent_wrapper(self.env.cr)
+                    similarities = ["word_similarity(%s, %s)" % (unaccent("%(search)s"), unaccent(field)) for field in fields]
+                    similar = ["%s <%%%% %s" % (unaccent("%(search)s"), unaccent(field)) for field in fields]
+                else:
+                    similarities = ["word_similarity(%s, %s)" % ("%(search)s", field) for field in fields]
+                    similar = ["%s <%%%% %s" % ("%(search)s", field) for field in fields]
+                best_similarity_str = 'GREATEST(%s)' % ', '.join(similarities)
+
+                query = """
+                    SELECT id, {best_similarity_str} AS _best_similarity
+                    FROM {table}
+                    WHERE {where}
+                    ORDER BY _best_similarity desc
+                    LIMIT 1000
+                """.format(
+                    best_similarity_str=best_similarity_str,
+                    table=model._table,
+                    where=' OR '.join(similar),
+                )
```

Comment:
I'm really not enthused with these queries generated from concatenating strings which come from overridden functions from random modules, this feels like it's opening doors to a host of future SQL injection vectors. The input fields list is *currently* filtered based on the model, but that feels like an easy thing to lose in a future error.

I think this should use `psycopg2.sql` wrappers throughout, the fields should be as `sql.Identifier` at definition site, and `basic_enumerate_words` should strip out the identifier-ness (just access the `string` property of the identifier).

This way if a non-`Composable` sneaks in somehow it should trigger an error.

The only annoyance is that `get_unaccent_wrapper` does not understand or output `psycopg2.sql` but we could probably just have a version which works in those terms?

@odony thoughts?

Reply:
yes please, I find the `psycopg2.sql` wrappers horrible to use due to the increased verbosity, but for that level of raw SQL crafting it's definitely worth it.

For the unaccent thing, we could either support `psycopg2.sql` there, or have a few explicit checks on `registry.has_unaccent`, whichever yields the simplest/more readable code.

On a related note, `escape_psql()` should never be used for crafting raw SQL (as it is both inferior and redundant with `pyscopg2.sql`). Its uses should be limited to preprocessing filter criterions for use with `[=]like`-class operators in domain terms, wherever appropriate.

Reply:
What about building the query relying on private methods of `model` - then build the "greatest similarity" expression from the obtained field aliases ?

Reply:
Converted to psycopg2.sql